src/lidar_subscriber.cpp

```diff
@@ -0,0 +1,92 @@
+#include <memory>
+#include <cmath>
+#include <queue>
+#include "rclcpp/rclcpp.hpp"
+#include "sensor_msgs/msg/laser_scan.hpp"
+#include "core-structs.h"
+#include <limits>
+constexpr float LANDMARK_THRESHOLD_m = 0.005;
+
+bool same_landmark(Observation2D first, Observation2D second){
+  float r1 = first.range_m;
+  float r2 = second.range_m;
+  float theta1 = first.bearing_rad;
+  float theta2 = second.bearing_rad;
+  float distance = sqrt(pow(r1*sin(theta1)-r2*sin(theta2),2.0)+pow(r1*cos(theta1)-r2*cos(theta2), 2.0));
+  return  distance <= LANDMARK_THRESHOLD_m;
+}
+
+Observation2D merge(Observation2D existing_landmark, Observation2D new_measurement, int count){
+  struct Observation2D average;
+  average.range_m = ((count-1)*existing_landmark.range_m+new_measurement.range_m)/count;
+  average.bearing_rad = ((count-1)*existing_landmark.bearing_rad+new_measurement.bearing_rad)/count;
+  return average;
+}
+
+Observation2D merge(Observation2D first, Observation2D second, int count1, int count2){
+  //standard weighted average --> between first and last landmark
+  struct Observation2D average;
+  average.range_m = first.range_m*(count1/(count1+count2))+second.range_m*(count2/(count1+count2));
+  average.bearing_rad = first.bearing_rad*(count1/(count1+count2))+second.bearing_rad*(count2/(count1+count2));
+  return first;
+}
+
+class LidarSubscriber : public rclcpp::Node
+{
+  public:
+    LidarSubscriber()
+    : Node("lidar_subscriber")
+    {
+      subscription = this->create_subscription<sensor_msgs::msg::LaserScan>(
+      "scan", 10, std::bind(&LidarSubscriber::laserscan_to_landmarks, this, std::placeholders::_1));
+    }
+  private:
+    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr subscription;
+    std::queue<Observation2D> laserscan_to_landmarks(const sensor_msgs::msg::LaserScan::SharedPtr msg){
+        std::vector<float> ranges = msg->ranges;
```

Comment:
Be careful here: `ranges` is a c-style array, so it can't be implicitly converted to a `std::vector`. See [this page](https://stackoverflow.com/questions/8777603/what-is-the-simplest-way-to-convert-array-to-vector) for more info.

Reply:
I think I will have to loop through and add each value to the struct one by one :( in order to initialize it as a c style array. Instead I'll just use msg->ranges[i] where I previously had ranges[i]

Reply:
Good thinking. Though I would be careful here, with empty arrays or mis-calculated array sizes. Is there a way to definitively know the array size? If not, we should add in some error handling code here to prevent de-referencing nullptrs or accessing out-of-bound entries.